### PR TITLE
SONARJAVA-1298 type inference for resolution of methods using type variable

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -78,13 +78,16 @@ public Scope createStaticStarImportScope(JavaSymbol owner) {
...
 
   public JavaType resolveTypeSubstitution(JavaType type, JavaType definition) {
-    if(isParametrizedType(definition)) {
+    if (isParametrizedType(definition)) {
       return substituteTypeParameter(type, ((JavaType.ParametrizedTypeJavaType) definition).typeSubstitution);
     }
     return type;
   }
 
   private JavaType substituteTypeParameter(JavaType type, TypeSubstitution substitution) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "substituteTypeParameter" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 2]
```diff
@@ -499,6 +517,91 @@ private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, J
...
+          // we found all the substitution
+          break;
+        }
+      }
+    }
+    return typeSubstitution;
+  }
+
+  // FIXME workaround to ignore bounds in type variable - WIP
+  private boolean hasBounds(List<JavaType.TypeVariableJavaType> typeVariablesToInfer) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Make "hasBounds" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 3]
```diff
@@ -499,6 +517,91 @@ private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, J
...
+    if (formalType.isTagged(JavaType.TYPEVAR) && currentSubstitution.substitutedType(formalType) == null) {
+      JavaType expectedType = argType;
+      if (expectedType.isPrimitive()) {
+        expectedType = expectedType.primitiveWrapperType;
+      }
+      currentSubstitution.add((JavaType.TypeVariableJavaType) formalType, expectedType);
+    }
+  }
+
+  private TypeSubstitution mergeTypeSubstitutions(TypeSubstitution currentSubstitution, TypeSubstitution newSubstitution) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Make "mergeTypeSubstitutions" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 4]
```diff
@@ -512,15 +615,11 @@ private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, J
...
-  private static boolean usesTypeParameter(JavaSymbol.MethodJavaSymbol methodSymbol) {
-    return !methodSymbol.typeVariableTypes.isEmpty();
-  }
-
   /**
    * @param argTypes types of arguments
    * @param formals  types of formal parameters of method
    */
-  private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> formals, boolean isVarArgs, boolean autoboxing, boolean isParameterized) {
+  private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> formals, boolean isVarArgs, boolean autoboxing) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "isArgumentsAcceptable" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 5]
```diff
@@ -499,6 +517,91 @@ private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, J
     return mostSpecific;
   }
 
+  private TypeSubstitution inferTypeSubstitution(List<JavaType.TypeVariableJavaType> typeVariableTypes, List<JavaType> formals, List<JavaType> argTypes) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: name is absolutely meaningless

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 6]
```diff
@@ -78,13 +78,16 @@ public Scope createStaticStarImportScope(JavaSymbol owner) {
...
   public JavaType resolveTypeSubstitution(JavaType type, JavaType definition) {
-    if(isParametrizedType(definition)) {
+    if (isParametrizedType(definition)) {
       return substituteTypeParameter(type, ((JavaType.ParametrizedTypeJavaType) definition).typeSubstitution);
     }
     return type;
   }
 
   private JavaType substituteTypeParameter(JavaType type, TypeSubstitution substitution) {
+    if (type == null) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: huh ?

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: Why this can be null  ? if it can why it is not annotated ?

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 7]
```diff
@@ -97,13 +100,23 @@ private JavaType substituteTypeParameter(JavaType type, TypeSubstitution substit
...
+          substitutedType = ((JavaType.WildCardType) substitutedType).bound;
+        }
         return parametrizedTypeCache.getWildcardType(substitutedType, wildCardType.boundType);
       }
     }
+    if (type.isTagged(JavaType.ARRAY)) {
+      JavaType.ArrayJavaType arrayType = (JavaType.ArrayJavaType) type;
+      substitutedType = substituteTypeParameter(arrayType.elementType, substitution);
+      if (substitutedType != null) {
+        return new JavaType.ArrayJavaType(substitutedType, symbols.arrayClass);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: Ouch

**Comment by m-g-sonar** [comment]
Will be reworked in other branch.

**Pull request closed without merging**